python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,471 @@
+        #with self.thread_lock:
+        self.shared_testers_dag[testers[-1].getTestDir()] = tester_dag
+
+        # Set a pending status on these jobs _before_ launching them
+        for job in ind_jobs:
+            self.scheduled_name_to_object[job].setStatus('QUEUED', self.scheduled_name_to_object[job].bucket_pending)
+
+        # Launch the jobs
+        for job in ind_jobs:
+            self.runnerGo(job)
+
+    # Launch jobs stored in our queues. Method is blocked until both queue(s) and active jobs are empty
+    def waitFinish(self):
+        while len(self.active_jobs) != 0 or not self.runner_queue.empty() or not self.status_queue.empty():
+            # Using a thread lock, check and do things every iteration
+            with self.thread_lock:
+                self.handleActiveTests()
+
+            # sleep for just a tick or two
+            sleep(0.1)
+
+        ### Runner Pool
+        # Wait for the runner pool to empty before waiting on the status pool
+        # The runner pool and status pool can exhange jobs between each other but it is always the status pool
+        # that will have items in the queue last (that last job with a finished status to be printed)
+
+        # Close the runner_pool so jobs can no longer be scheduled for launching
+        self.runner_pool.close()
+
+        # Wait for the runner_pool to empty
+        self.runner_pool.join()
+
+        # Close the status_pool so jobs can no longer be scheduled for status printing
+        self.status_pool.close()
+
+        # Wait for the status_pool to empty
+        self.status_pool.join()
+
+    # Method to do things needed to actively running tests.
+    # NOTE: We are in a thread locked state, so be quick
+    def handleActiveTests(self):
+        # loop through all the active tests and do something
+        for tester_name in self.active_jobs:
+            tester = self.scheduled_name_to_object[tester_name]
+
+            # Handle long running tests
+            self.doLongRunningTests(tester)
+
+            # Handle timeouts
+            self.doTimeoutTests(tester)
+
+    def doLongRunningTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            min_time = self.default_reported_time
+            if tester.specs.isValid('min_reported_time'):
+                min_time = float(tester.getMinReportTime())
+
+            if clock() - tester.getStartTime() > min_time:
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('RUNNING...', tester.bucket_pending)
+                self.harness.handleTestStatus(tester)
+
+    def doTimeoutTests(self, tester):
+        if tester.getTestName() not in self.reported_jobs:
+            if clock() - tester.getStartTime() > float(tester.getMaxTime()):
+                self.reported_jobs.add(tester.getTestName())
+                tester.setStatus('TIMEOUT', tester.bucket_fail)
+                self.harness.handleTestStatus(tester)
```

Comment:
Lock here with a callback to an object that doesn't even know about threading. Keep your locks very local, right where you need them.